pkg/genericapiserver/enablement_config.go

```diff
@@ -0,0 +1,159 @@
+	// unusual, but we need it for compatibility with old code for now.  An empty set means
+	// enable all, a non-empty set means that all other resources are disabled.
+	EnabledResources sets.String
+}
+
+var _ APIResourceEnablement = &APIResourceEnablementConfig{}
+
+// AnyVersion is used during resource enablement checks to see if any version of the resource is enabled
+const AnyVersion = "."
+
+type APIResourceEnablementConfig struct {
+	GroupVersionDisablers map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig
+}
+
+func NewAPIResourceEnablementConfig() *APIResourceEnablementConfig {
+	return &APIResourceEnablementConfig{GroupVersionDisablers: map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig{}}
+}
+
+func NewAPIGroupVersionEnablementConfig() *APIGroupVersionEnablementConfig {
+	return &APIGroupVersionEnablementConfig{Enable: true, DisabledResources: sets.String{}, EnabledResources: sets.String{}}
+}
+
+func (o *APIResourceEnablementConfig) DisableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = false
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = true
+	}
+}
+
+func (o *APIResourceEnablementConfig) DisableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].DisabledResources.Insert(resource.Resource)
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].EnabledResources.Insert(resource.Resource)
+		o.GroupVersionDisablers[version].DisabledResources.Delete(resource.Resource)
+	}
+}
+
+func (o *APIResourceEnablementConfig) ResourceEnabled(resource unversioned.GroupVersionResource) bool {
+	if resource.Version == AnyVersion {
+		return o.anyVersionOfResourceEnabled(resource)
+	}
+
+	return o.exactResourceEnabled(resource)
+}
+
+func (o *APIResourceEnablementConfig) anyVersionOfResourceEnabled(resource unversioned.GroupVersionResource) bool {
```

Comment:
I guess I'm saying to document.

On Tue, Mar 22, 2016 at 1:32 PM, David Eads <notifications@github.com>
wrote:

> In pkg/genericapiserver/enablement_config.go
> <https://github.com/kubernetes/kubernetes/pull/23067#discussion_r57063490>
> :
>
> > +		}
> > +
> > +		o.GroupVersionDisablers[version].EnabledResources.Insert(resource.Resource)
> > +		o.GroupVersionDisablers[version].DisabledResources.Delete(resource.Resource)
> > +	}
> > +}
> > +
> > +func (o *APIResourceEnablementConfig) ResourceEnabled(resource unversioned.GroupVersionResource) bool {
> > +	if resource.Version == AnyVersion {
> > +		return o.anyVersionOfResourceEnabled(resource)
> > +	}
> > +
> > +	return o.exactResourceEnabled(resource)
> > +}
> > +
> > +func (o *APIResourceEnablementConfig) anyVersionOfResourceEnabled(resource unversioned.GroupVersionResource) bool {
>
> Pretty sure this wouldn't work right across resource renames.
>
> It doesn't take the rename into account. If you care about renames, you'll
> need to specific and request exact matches.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23067/files/cdef546dee0c2d222b6ca0cfee281079b5be2ce6#r57063490>
>